hotham/src/shaders/pbr.frag

```diff
@@ -84,33 +92,28 @@ vec3 getNormal(uint normalTextureID)
 	return normalize(TBN * tangentNormal);
 }
 
-// Add some fake IBL. Texture reads are PHENOMENALLY expensive on mobile so we'll need to come up with a
-// faster solution than reading all the IBL images.
+// Calculation of the lighting contribution from an optional Image Based Light source.
 vec3 getIBLContribution(PBRInfo pbrInputs, vec3 n, vec3 reflection)
 {
-	vec4 irradiance  = vec4(0.5);
-	vec4 ibl_diffuse = irradiance * vec4(pbrInputs.diffuseColor, 1.0);
-	return tonemap(ibl_diffuse).xyz;
-}
+	// Flip the y axis of the reflection vector; otherwise our cubemap will be upside down.
+	// TODO: Fix this by pre-flipping the cubemaps.
+	reflection.y *= -1.;
+	float lod = pbrInputs.perceptualRoughness * float(DEFAULT_CUBE_MIPMAP_LEVELS -1);
 
-// TODO: RESTORE
-// Calculation of the lighting contribution from an optional Image Based Light source.
-// Precomputed Environment Maps are required uniform inputs and are computed as outlined in [1].
-// See our README.md on Environment Maps [3] for additional discussion.
-// vec3 getIBLContributionWithTextures(PBRInfo pbrInputs, vec3 n, vec3 reflection)
-// {
-// 	float lod = (pbrInputs.perceptualRoughness * DEFAULT_CUBE_MIPMAP_LEVELS);
-// 	// retrieve a scale and bias to F0. See [1], Figure 3
-// 	vec3 brdf = (texture(textures[BRDF_LUT_TEXTURE_ID], vec2(pbrInputs.NdotV, 1.0 - pbrInputs.perceptualRoughness))).rgb;
-// 	// vec3 diffuseLight = tonemap(texture(samplerIrradiance, n)).rgb;
-// 	// vec3 specularLight = tonemap(textureLod(prefilteredMap, reflection, lod)).rgb;
-
-// 	// vec3 diffuse = diffuseLight * pbrInputs.diffuseColor;
-// 	// vec3 specular = specularLight * (pbrInputs.specularColor * brdf.x + brdf.y);
-
-// 	// return diffuse + specular;
-// 	return vec3(1.);
-// }
+	// retrieve a scale and bias to F0. See [1], Figure 3
+	vec2 brdfSamplePoint = clamp(vec2(pbrInputs.NdotV, 1.0 - pbrInputs.perceptualRoughness), vec2(0.0, 0.0), vec2(1.0, 1.0));
+	vec3 brdf = texture(textures[BRDF_LUT_TEXTURE_ID], brdfSamplePoint).rgb;
+
+	// Get diffuse and specular light values
+	vec3 diffuseLight = tonemap(texture(cubeTextures[SAMPLER_IRRADIANCE_TEXTURE_ID], n)).rgb;
+	vec3 specularLight = tonemap(textureLod(cubeTextures[ENVIRONMENT_MAP_TEXTURE_ID], reflection, lod)).rgb;
```

Comment:
If you tonemap here, it does not make sense to blend it later on because tonemapping is non-linear.

Reply:
Yep, agreed